processor/probabilisticsamplerprocessor/README.md

```diff
@@ -15,51 +15,122 @@
+`sampling_priority`.
+
+## Sampling algorithm
+
+### Hash seed
+
+The hash seed method uses the FNV hash function applied to either a
+Trace ID (spans, log records), or to the value of a specified
+attribute (only logs).  The hashed value, presumed to be random, is
+compared against a threshold value that corresponds with the sampling
+percentage.
+
+This mode requires configuring the `hash_seed` field.  This mode is
+enabled when the `hash_seed` field is not zero, or when log records
+are sampled with `attribute_source` is set to `record`.
+
+In order for hashing to be consistent, all collectors for a given tier
+(e.g. behind the same load balancer) must have the same
+`hash_seed`. It is also possible to leverage a different `hash_seed`
+at different collector tiers to support additional sampling
+requirements.
+
+This mode uses 14 bits of sampling precision.
+
+### Error handling
+
+This processor considers it an error when the arriving data has no
+randomess.  This includes conditions where the TraceID field is
+invalid (16 zero bytes) and where the log record attribute source has
+zero bytes of information.
+
+By default, when there are errors determining sampling-related
+information from an item of telemetry, the data will be refused.  This
+behavior can be changed by setting the `fail_closed` property to
+false, in which case erroneous data will pass through the processor.
+
+## Configuration
 
 The following configuration options can be modified:
-- `hash_seed` (no default, optional): An integer used to compute the hash algorithm. Note that all collectors for a given tier (e.g. behind the same load balancer) should have the same hash_seed.
-- `sampling_percentage` (required): Percentage at which logs are sampled; >= 100 samples all logs, 0 rejects all logs.
-- `attribute_source` (default = traceID, optional): defines where to look for the attribute in from_attribute. The allowed values are `traceID` or `record`.
-- `from_attribute` (default = null, optional): The optional name of a log record attribute used for sampling purposes, such as a unique log record ID. The value of the attribute is only used if the trace ID is absent or if `attribute_source` is set to `record`.
-- `sampling_priority` (default = null, optional): The optional name of a log record attribute used to set a different sampling priority from the `sampling_percentage` setting. 0 means to never sample the log record, and >= 100 means to always sample the log record.
 
-## Hashing
+- `sampling_percentage` (32-bit floating point, required): Percentage at which items are sampled; >= 100 samples all items, 0 rejects all items.
+- `hash_seed` (32-bit unsigned integer, optional, default = 0): An integer used to compute the hash algorithm. Note that all collectors for a given tier (e.g. behind the same load balancer) should have the same hash_seed.
+- `fail_closed` (boolean, optional, default = true): Whether to reject items with sampling-related errors.
```

Comment:
The current implementation has an odd inconsistency which this is meant to resolve in a user-configurable way. In cases where there is no randomness, the sampling decision is fixed and therefore depends on the configured probability. For some probabilities, the failure would be open, and for other probabilities, the failure would be closed--depending on the hash seed.

I added this test to demonstrate:

```
func TestEmptyHashFunction(t *testing.T) {
	// With zero bytes of randomness, seed=0:
	hashed32 := computeHash([]byte{}, 0)
	hashed := uint64(hashed32 & bitMaskHashBuckets)
	require.Equal(t, uint64(0x3515), hashed)
	require.InDelta(t, 0.829, float64(hashed)/float64(numHashBuckets), 0.001)

	// With 16 bytes of 0s, seed=0:
	var b [16]byte
	hashed32 = computeHash(b[:], 0)
	hashed = uint64(hashed32 & bitMaskHashBuckets)
	require.Equal(t, uint64(0x2455), hashed)
	require.InDelta(t, 0.568, float64(hashed)/float64(numHashBuckets), 0.001)
}
```

So for empty TraceIDs (16 bytes of 0s) are sampled at 56.8% and above, and logs with missing attribute values (0 bytes) are sampled at 82.9% and above. Since I expect most users are configuring probabilities at 50% or below, most users would never see these items of telemetry and never knew there was a problem. I think reversing the decision w/ FailClosed=false by default, so that users receive these spans, is a better outcome.

Reply:
Sounds good!